backend/common/repo.py

```diff
@@ -0,0 +1,232 @@
+
+                except ValueError:
+                    pass
+
+        return self._release
+
+    def get_release_index(self) -> typing.Dict[str, "DpkgRepo.ReleaseEntry"]:
+        """
+        Find and return contents of Release file.
+
+        :raises GeneralRepoException if the Release file cannot be found.
+        :return: string
+        """
+        resp = requests.get(self._get_parent_url(self._url, 2, "Release"))
+        try:
+            self._flat = resp.status_code in [http.HTTPStatus.NOT_FOUND, http.HTTPStatus.FORBIDDEN]
+            self._flat_checked = 1
+            self._release = self._parse_release_index(resp.content.decode("utf-8"))
+            if resp.status_code not in [http.HTTPStatus.NOT_FOUND, http.HTTPStatus.OK, http.HTTPStatus.FORBIDDEN]:
+                raise GeneralRepoException("HTTP error {} occurred while connecting to the URL".format(resp.status_code))
+
+            if not self._release and self.is_flat():
+                resp = requests.get(self._get_parent_url(self._url, 0, "Release"))
+                if resp.status_code == http.HTTPStatus.OK:
+                    self._release = self._parse_release_index(resp.content.decode("utf-8"))
+        finally:
+            resp.close()
+
+        if not self._release:
+            raise GeneralRepoException("Repository seems either broken or has unsupported format")
+
+        return self._release
+
+    @staticmethod
+    def _get_parent_url(url, depth=1, add_path=""):
+        """
+        Get parent url from the given one.
+
+        :param url: an url
+        :return: parent url
+        """
+        p_url = parse.urlparse(url)
+        p_path = p_url.path.rstrip("/").split("/")
+        if depth:
+            p_path = p_path[:-depth]
+
+        return parse.urlunparse(parse.ParseResult(scheme=p_url.scheme, netloc=p_url.netloc,
+                                                  path="/".join(p_path + add_path.strip("/").split("/")) or "/",
+                                                  params=p_url.params, query=p_url.query, fragment=p_url.fragment))
+
+    def is_flat(self) -> bool:
+        """
+        Detect if the repository has flat format.
+
+        :return:
+        """
+        if self._flat_checked is None:
+            self.get_release_index()
+
+        return bool(self._flat)
+
+    def verify_packages_index(self) -> bool:
+        """
+        Verify Packages index against all listed checksum algorithms.
+
+        :param name: name of the packages index
+        :return: result (boolean)
+        """
+        name, data = self.get_pkg_index_raw()
+        entry = self.get_release_index().get(name)
+        for algorithm in ["md5", "sha1", "sha256"]:
+            if entry is None:
+                res = False
+            else:
+                res = getattr(hashlib, algorithm)(data).hexdigest() == getattr(entry.checksum, algorithm)
+            if not res:
+                break
```

Comment:
Is this algorithm really correct? 

If the entry do not specify one of the 3, it will fail even if the other checksums match.
For me it looks like this require that all 3 checksums are available.

Not sure if debian defines it that way. But in that case I would just check `sha256` and stop.

Reply:
The point is to verify all of them. Yes, it is correct. :wink:

Reply:
Is it guaranteed that all 3 must always exist?
Does old Debian distros support `sha256`?
Do new Debian distros still use `md5` ? Think about FIPS ...

Reply:
For Ubuntu they have three all the time. For FIPS, they simply don't use the md5 and sha1. Even is interesting that [they say](https://wiki.debian.org/DebianRepository/Format#MD5Sum.2C_SHA1.2C_SHA256) "must require 256 or 512", while I haven't seen yet some `Release` provides 512. The only concern here is that I found some ancient `Release` from Debian that doesn't include SHA1. AFAIK that is not the case for Ubuntu, but your point is heard: I will update this algorithm to check only those that are provided in the Release file.